PICO: Adding ADC

blckmn opened the pull request:
```resource ADC_BATT A40```

Tested as working with a 15.6v bench supply as source.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * RP2350 (Pico) ADC driver added with DMA-backed continuous sampling and adcGetValue API.

* **Refactor**
  * ADC API unified to a source-centric model and propagated across telemetry, sensors and drivers; internal calibration preserved.

* **Chores**
  * Buildfiles and platform traits updated; CLI ADC parameter now platform-conditional; simulator/test ADC stubs removed.

* **Platform**
  * STM32/AT32 ADC backends consolidated with a common STM32 ADC implementation.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

blckmn commented:
Tested VBAT as working.

ledvinap commented on src/platform/PICO/adc_pico.c:
```diff
@@ -0,0 +1,208 @@
+        mask |= (1 << adcOperatingConfig[i].channel);
+    }
+
+    const uint8_t sources = BITCOUNT(mask);
```
```suggestion
    const unsigned sources = popcount(mask);
```

ledvinap commented on src/platform/PICO/adc_pico.c:
```diff
@@ -0,0 +1,208 @@
+            continue;
+        }
+
+        if (mask == 0) {
```
You can use `int firstEnabled = llog2(mask & -mask);` after loop

ledvinap commented on src/platform/PICO/adc_pico.c:
```diff
@@ -0,0 +1,208 @@
+        Sampling requires 96 cycles per sample
+        Sample as slow as possible
+    */
+    adc_set_clkdiv((65535.f + 255.f/256.f));
```
```suggestion
    adc_set_clkdiv(65535.f + 255.f / 256.f);
```

ledvinap commented on src/platform/PICO/adc_pico.c:
```diff
@@ -0,0 +1,208 @@
+
+uint16_t adcGetValue(adcSource_e source)
+{
+    if (!adcOperatingConfig[source].enabled) {
```
```suggestion
    if (source >= ARRAYLEN(adcOperatingConfig) || !adcOperatingConfig[source].enabled) {
```

blckmn replied:
Although source is an enum, and the arraylength is set by that enum count. Need to cast the enum as unsigned.

ledvinap commented on src/platform/common/stm32/adc_impl.c:
```diff
@@ -0,0 +1,104 @@
+        return false;
+    }
+
+    for (int map = 0 ; map < ADC_TAG_MAP_COUNT ; map++) {
```
maybe:
```suggestion
    for (int map = 0 ; map < ARRAYLEN(adcTagMap); map++) {
```
Or you can iterate entries directly:
```suggestion
    for (const adcTagMap_t m = adcTagMap; m < ARRAYEND(adcTagMap); m++) {
```

blckmn commented on src/main/pg/adc.h:
```diff
@@ -23,11 +23,17 @@
 #include "drivers/dma_reqmap.h"
 
+#ifdef USE_DMA_SPEC
+#include "platform/adc_impl.h"
```
Agree... will think about it for another PR.

mjs1441 commented on src/platform/STM32/include/platform/platform.h:
```diff
@@ -478,14 +478,14 @@ extern uint8_t _dmaram_end__;
 
 #define PLATFORM_TRAIT_ADC_INTERNAL 1
```
USE_ADC_INTERNAL ?

mjs1441 commented:
Is it written somewhere what the different traits and definitions, such as PLATFORM_TRAIT_ADC*, signify? (My comments on the PR might be misguided)

mjs1441 commented:
Thanks @ledvinap .

Still a bit to do to enable ADC_INTERNAL (or make use of new PLATFORM_TRAIT_ADC_INTERNAL) and temperature readings on PICO, but maybe that's for another PR.

blckmn commented:
@mjs1441 good pickup. the PLATFORM_TRAIT_ADC_INTERNAL is not actually used (was planned to be). USE_ADC_INTERNAL is now usable from config.h - and is present already and utilised in prior releases. PLATFORM_TRAIT_ADC_DEVICE is actually used, and represents the need to specify the device. It is possible to remove also given the implementation of adcOperatingConfig is now in platform and not in the main.

mjs1441 commented on src/platform/PICO/target/RP2350B/target.h:
```diff
@@ -61,6 +61,13 @@
+
+#undef USE_SOFTSERIAL1
+#undef USE_SOFTSERIAL2
+
```
Looks like you can drop changes in RP2350B/target.h vs master (merge made some duplicate lines)

mjs1441 replied:
actually, I think we lost USE_ADC

mjs1441 commented on src/platform/PICO/adc_pico.c:
```diff
@@ -0,0 +1,219 @@
+    while (!adc_fifo_is_empty()) {
+        uint16_t result = adc_fifo_get();
+        // The ADC hardware tells us which channel the sample is from.
+        uint8_t channel = result >> 12;  // Extract channel from top 4 bits
```
It turns out that, unfortunately, the FIFO does not record any channel information!

So, we could keep track of what we think the channel is, in a variable that we move along each time we call adc_fifo_get.  It's a bit fragile. There's a risk it can get out of sync, if the FIFO overflows (probably unlikely for us at low sample rates).

Maybe better (more solidly reliable) is to trigger the FIFO interrupt on a single conversion, then in the handler, set a timed callback (hardware_alarm_set_callback etc.) to trigger the next conversion on the known next channel?

blckmn replied:
Having the channel information would be an absolute bonus here, and would make the free running round robin and dma ring buffer absolutely the best option for continuously sampling the ADC. Alas. Will provide the feed back to RP.

mjs1441 commented on src/platform/PICO/adc_pico.c:
```diff
@@ -0,0 +1,236 @@
+    // The size is 2^X = Z half-words
+    // note ADC handles the reset, the ring is never triggered
+    const unsigned sizeBits = 32 - __builtin_clz(ARRAYLEN(adcValues) - 1);
+    channel_config_set_ring(&cfg, true, sizeBits);
```
I think we would want
 enabled channels = 2 -> sizeBits = 1
 enabled channels = 4 -> sizeBits = 2
 enabled channels = 8 -> sizeBits = 3
but I don't see how it would work for any other number of enabled channels (the dma would put values from say channel 0 in different slots each time it loops).

mjs1441 replied:
The latest pico-sdk has convenience functions for programming the DMA to do a set number of transfers and self-trigger, which I guess is what we want. 

See ```dma_encode_transfer_count_with_self_trigger(uint transfer_count)```
in https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/hardware_dma/include/hardware/dma.h

blckmn replied:
I will look into that - briefly didn't see where it resets the write_addr so not sure it will work as we planned.

ledvinap commented on src/platform/PICO/include/platform/platform.h:
```diff
@@ -131,3 +131,16 @@ extern uint32_t systemUniqueId[3];
+
+// TODO: identify correct values for RP2350 internal temp sensor, or move conversion into adc_pico.c.
+#if defined(USE_ADC_INTERNAL)
+#error "PICO: Internal temp/Vref calibration constants are not verified for RP2350. Provide correct values or implement conversion in adc_pico.c."
```
Maybe use just warning, it will trigger -Werror in normal build anyway
```suggestion
#warning "PICO: Internal temp/Vref calibration constants are not verified for RP2350. Provide correct values or implement conversion in adc_pico.c."
```

ledvinap commented on src/platform/PICO/include/platform/platform.h:
```diff
@@ -131,3 +131,16 @@ extern uint32_t systemUniqueId[3];
+#error "PICO: Internal temp/Vref calibration constants are not verified for RP2350. Provide correct values or implement conversion in adc_pico.c."
+// Placeholder values below only as guidance; do not ship with these.
+#define VREFINT_CAL_VREF                   ( 3300U)           /* Analog voltage reference (Vref+) value with which temperature sensor has been calibrated in production (tolerance: +-10 mV) (unit: mV). */
+#define TEMPSENSOR_CAL1_TEMP               (( int32_t)   30)  /* Internal temperature sensor, temperature at which temperature sensor has been calibrated in production for data into TEMPSENSOR_CAL1_ADDR (tolerance: +-5 DegC) (unit: DegC). */
```
```suggestion
#define TEMPSENSOR_CAL1_TEMP               27  /* Internal temperature sensor, reference temperature from datasheet (unit: DegC). */
```

ledvinap commented on src/platform/PICO/include/platform/platform.h:
```diff
@@ -131,3 +131,16 @@ extern uint32_t systemUniqueId[3];
+#if defined(USE_ADC_INTERNAL)
+#error "PICO: Internal temp/Vref calibration constants are not verified for RP2350. Provide correct values or implement conversion in adc_pico.c."
+// Placeholder values below only as guidance; do not ship with these.
+#define VREFINT_CAL_VREF                   ( 3300U)           /* Analog voltage reference (Vref+) value with which temperature sensor has been calibrated in production (tolerance: +-10 mV) (unit: mV). */
```
On stm32, this is reference value for VREF calibration,  TEMPSENSOR is related to temp sensor calibration.

It would be best to move adcInternalComputeTemperature to platform and use datasheet equation for PICO (but temp sensor would be quite useless without calibration)

blckmn replied:
Will do that in another PR when fully setting up the internal channel.

blckmn commented on src/platform/PICO/adc_pico.c:
```diff
@@ -0,0 +1,236 @@
+        true,  // Write each completed conversion to the sample FIFO
+        true,  // Enable DMA data request (DREQ)
+        1,     // Fire DREQ when FIFO has at least 1 sample
+        false, // We won't see the ERR bit because of 12-bit reads; disable
```
I would have preferred the channel information in those 4 bits... would have seemed more logica, and would have enabled round robin and dma ring buffer. I will provide the feedback to RP.

blckmn commented on src/platform/PICO/adc_pico.c:
```diff
@@ -0,0 +1,236 @@
+    // Set ring buffer behavior: after writing the last element, wrap back to the start.
+    // The size is 2^X = Z half-words
+    // note ADC handles the reset, the ring is never triggered
+    const unsigned sizeBits = 32 - __builtin_clz(ARRAYLEN(adcValues) - 1);
```
This was a work in progress. The issue is the lack of channel information in the output from the round robin.

Now the code is allowing for the FIFO to overflow if the interrupt is late. The interrupt simply halts the ADC free running, 

This provides a consistently positioned output buffer order from the DMA (and no need for ring buffer).

blckmn commented on src/platform/PICO/adc_pico.c:
```diff
@@ -0,0 +1,236 @@
+                break;
+            }
+
+            const int pin = DEFIO_TAG_PIN(adcOperatingConfig[i].tag);
```
I will fix that up in another PR - as it will impact other files also.

ledvinap commented:
@blckmn : Can you avoid force-pushing? A lot of of github features (for example seeing changes since last review) do not work. And at least in past come code comments got lost

ledvinap commented on src/platform/PICO/adc_pico.c:
```diff
@@ -0,0 +1,236 @@
+{
+    adc_init();
+
+    memset(adcOperatingConfig, 0, sizeof(adcOperatingConfig));
```
> Odd. I was always thought:
> 
> * Passing arr or &arr[0] is equivalent in this context, due to the decay anyway.

it is. And it would be correct for `int fn(adcOperatingConfig_t *cfg, size_t count)`

> * Do not pass &arr (address of the whole array), as it has a different type and may lead to incorrect behavior.

IMO for memset the arguments are object address + its size. So in this case you want pointer to whole array and not first element. Also it is consistent with non-array case `int i; memset(&i, 0, sizeof(i));``

blckmn commented:
> @blckmn : Can you avoid force-pushing? A lot of of github features (for example seeing changes since last review) do not work. And at least in past come code comments got lost

Force pushing is required for squashing, and if one wants to reset back. Ultimately the code gets entirely squashed also. But I get the see changes since last review... in future I will set the PR in draft, so as not to pre-empt a review prior to completion. I was testing various techniques with the RP2350 to see what works, what was fragile etc given the ADC deficiencies.

ledvinap commented:
> Force pushing is required for squashing, and if one wants to reset back. Ultimately the code gets entirely squashed also on merge. But I get the see changes since last review... in future I will set the PR in draft, so as not to pre-empt a review prior to completion. I was testing various techniques with the RP2350 to see what works, what was fragile etc given the ADC deficiencies.

It's not a big problem, just don't force-push when it is easy to avoid it, please. (you can use `git restore --source <commit> --worktree .` instead of reset)

ledvinap commented on src/main/drivers/adc.h:
```diff
@@ -26,70 +26,36 @@
+    ADC_RSSI,
+#ifdef USE_ADC_INTERNAL
+    // For certain processors internal sensors are treated in the similar fashion as regular ADC inputs
+    ADC_SOURCE_INTERNAL_FIRST_ID = 4,
```
```suggestion
    ADC_SOURCE_INTERNAL_FIRST_ID,
```

ledvinap commented on src/main/drivers/adc.h:
```diff
@@ -26,70 +26,36 @@
+uint16_t adcGetValue(adcSource_e source);
 
 #ifdef USE_ADC_INTERNAL
+extern int32_t adcVREFINTCAL; // ADC value (12-bit) of band gap with Vref = VREFINTCAL_VREF
+extern int32_t adcTSCAL1;
+extern int32_t adcTSCAL2;
+extern int32_t adcTSSlopeK;
```
(these are STM32 specific)

ledvinap commented on src/platform/PICO/adc_pico.c:
```diff
@@ -0,0 +1,275 @@
+    UNUSED(descriptor);
+
+    /*
+        Simply halt ADC as it will now (or soon to be) over flowing the FIFO
```
(it would take ages for fifo to overflow if 10Hz sampling clock was used)

ledvinap commented on src/platform/PICO/adc_pico.c:
```diff
@@ -0,0 +1,275 @@
+        Sample rate required at 10hz for X channels = 10 x X
+
+        65535.f + 255.f / 256.f for as slow as possible
+        0 for as fast as possible
```
```suggestion
        0 .. 96 for as fast as possible (immediately after 96 cycles of conversion)
```

ledvinap commented on src/platform/PICO/adc_pico.c:
```diff
@@ -0,0 +1,275 @@
+    channel_config_set_write_increment(&cfg, true); // Write to sequential addresses into adcValues
+    channel_config_set_dreq(&cfg, DREQ_ADC);
+
+    dma_channel_configure(dmaChannel, &cfg, adcValues, &adc_hw->fifo, 1, false);
```
Isn't transfer count number of transfers (channelCount)?

ledvinap commented on src/platform/PICO/adc_pico.c:
```diff
@@ -0,0 +1,275 @@
+    }
+
+    if (!is_adc_busy()) {
+        adcStart();
```
Low sampling clock would be better in this mode

blckmn commented:
@ledvinap any further thoughts?

ledvinap commented on src/main/pg/adc.c:
```diff
@@ -37,68 +37,28 @@ PG_REGISTER_WITH_RESET_FN(adcConfig_t, adcConfig, PG_ADC_CONFIG, 0);
 
 void pgResetFn_adcConfig(adcConfig_t *adcConfig)
```
I'm not sure about this. Ok for now (with better name), but we should find something better long-term

ledvinap commented on src/platform/PICO/adc_pico.c:
```diff
@@ -0,0 +1,260 @@
+} adcOperatingConfig_t;
+
+static adcOperatingConfig_t adcOperatingConfig[ADC_SOURCE_COUNT];
+static volatile uint16_t adcValues[PICO_ADC_CHANNEL_COUNT];
```
PICO_ADC_MAX_CHANNELS for now?

ledvinap commented on src/platform/PICO/adc_pico.c:
```diff
@@ -0,0 +1,260 @@
+        adcOperatingConfig[ADC_CURRENT].tag = config->current.ioTag;
+    }
+
+    uint32_t mask = 0u;
+
+    // loop over all possible channels and build the adcOperatingConfig to represent
+    // the set of enabled channels
+    for (int i = 0; i < ADC_SOURCE_COUNT; i++) {
+        if (!adcOperatingConfig[i].tag || i > PICO_ADC_MAX_CHANNELS) {
+            continue;
+        }
+
+        const int pin = DEFIO_TAG_PIN(adcOperatingConfig[i].tag);
+        const int channel = adcChannelByPin(pin);
+        if (channel >= 0) {
+            adcOperatingConfig[i].channel = channel;
+            adc_gpio_init(pin);
+            adcOperatingConfig[i].enabled = true;
+        }
+        mask |= (1u << channel);
+    }
+
+    // Map each enabled source to its DMA slot by channel rank (ascending channel order).
+    for (int i = 0; i < ADC_SOURCE_COUNT; i++) {
+        if (!adcOperatingConfig[i].enabled) {
+            continue;
+        }
+        const uint8_t ch = adcOperatingConfig[i].channel;
+        adcOperatingConfig[i].dmaIndex = (uint8_t)popcount(mask & ((1u << ch) - 1u));
+    }
+
+    const uint8_t channelCount = popcount(mask);
```
```suggestion

    // loop over all possible channels and build the adcOperatingConfig to represent
    // the set of enabled channels
    int channelsEnabled = 0;
    uint32_t mask = 0u;
    for (int i = 0; i < ADC_SOURCE_COUNT && channelsEnabled < PICO_ADC_MAX_CHANNELS; i++) {
        if (!adcOperatingConfig[i].tag) {
            continue;
        }

        const int pin = DEFIO_TAG_PIN(adcOperatingConfig[i].tag);
        const int channel = adcChannelByPin(pin);
        if (channel < 0) {
            continue;
        }
        adc_gpio_init(pin);
        adcOperatingConfig[i].enabled = true;
        adcOperatingConfig[i].channel = channel;
        adcOperatingConfig[i].dmaIndex = channelsEnabled;
        mask |= (1u << channel);
        channelsEnabled++;
    }
    const uint8_t channelCount = channelsEnabled;
```

(maybe use channelsEnabled in rest of function)

blckmn replied:
The order of the mask - dictated by the channel number (not the enum position) is what drives the dmaIndex - not the enum position. Therefore if the pins are not in order the channels will not be in order, and so one needs to evaluate the mask - not the enum.

ledvinap commented on src/platform/PICO/adc_pico.c:
```diff
@@ -0,0 +1,260 @@
+        65535.f + 255.f / 256.f for as slow as possible
+        0 for as fast as possible
+    */
+    adc_set_clkdiv(65535.f + 255.f / 256.f);
```
maybe just adc_set_clkdiv(65535); - it will sample every 65536 cycles. Using fractional part is .. strange (one of 256 conversion would be faster, 65537 is not power of 2)

But it will be better to set some predictable sampling frequency. Just handle padding correctly

blckmn replied:
The fractional component was simply to make it as slow as possible... there is only a 1 byte available for the fraction. But I agree it makes it look more complicated for marginal benefit.

ledvinap commented on src/platform/PICO/adc_pico.c:
```diff
@@ -0,0 +1,260 @@
+        We will never have 3 channels as 6 bytes is unavailable in the ring buffer,
+        hence the padding with the internal temp sensor to prevent misalignment.
+    */
+    const uint16_t sizeBits = llog2(popcount(mask)) * sizeof(uint16_t);
```
sizeBits i number of bits to wrap - number of bytes is 1<<size_bits
```suggestion
    const uint16_t sizeBits = llog2(popcount(mask)) + sizeof(uint16_t) / 8 - 1;
```

ledvinap commented on src/platform/common/stm32/adc_impl.c:
```diff
@@ -0,0 +1,164 @@
+{
+    adcGetChannelValues();
+
+#ifdef DEBUG_ADC_CHANNELS
```
(not related to this PR - DEBUG macro should be used)
